espresso/environment/4_confirmation_integrity_with_reorgs_test.go

```diff
@@ -0,0 +1,212 @@
+package environment_test
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	env "github.com/ethereum-optimism/optimism/espresso/environment"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/geth"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/e2esys"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/helpers"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"math/big"
+	"strconv"
+	"testing"
+	"time"
+)
+
+// Computes the hash of the content of a batch. Introduced for testing purposes only.
+// @param b batch
+// @return string containing the hash of a batch
+func BatchHash(b *derive.SingularBatch) string {
+
+	// Concatenate the transactions and other relevant metadata of the batch
+	str := ""
+
+	for _, tx := range b.Transactions {
+		str = str + tx.String()
+	}
+	str = str + b.EpochHash.String()
+	str = str + strconv.Itoa(int(b.Timestamp))
+	str = str + b.ParentHash.String()
+
+	h := sha256.New()
+	h.Write([]byte(str))
+	hash := h.Sum(nil)
+
+	res := hex.EncodeToString(hash)
+
+	return res
+}
+
+// This function computes a list where at least a few batches are sent to unfinalized L1 blocks.
```

Comment:
Nit: "sent to" is a bit confusing. Maybe "derived from"?

Reply:
I meant indeed that the batches (included in frames) are sent to L1 blocks that are unfinalized. The reason is that we want to check that after a L1 reorg (where the L1 unfinalized blocks are removed), the same batches will be posted again. So what about this new phrasing:

This function computes a list of batches where at least a few of them are included in frame transactions which are inserted into unfinalized L1 blocks.